### DISCUSSION TITLE
How to get the original text of an `AstNode`? `AbstractCstNode.text` returns `undefined`

### DISCUSSION THREAD

**Discussion by rpedrosanto** [eclipse-langium/langium#1028] [Q&A] [answered]
Hi,

I need to retrieve the original text of an `AstNode` from a successfully parsed langium document but when I use the [`text` getter](https://github.com/langium/langium/blob/0b603a61b4725bc74adf6c21adcff586e129e1e1/packages/langium/src/parser/cst-node-builder.ts#L126) I always get `undefined`. The same is true for every other `AbstractCstNode` getter. What would be the correct way to get that text?

Thanks in advance, I appreciate it. Best.

**Comment by msujew** [comment]
Hey @rpedrosanto,

Using the `text` getter is supposed to be the correct way to get the text from a parsed document. Can you share a reproducible sample? It works for me as intended:

![image](https://user-images.githubusercontent.com/4377073/232508630-476dfe1e-28f8-4d9c-ad9f-58fdec69ca64.png)

**Reply by rpedrosanto** [reply]
> Can you share a reproducible sample?

Not sure I could do that easily enough, but I can see a difference in what you posted, the prototype for you is `AbstractCstNode` and what I get when I console log it is `Object`, maybe the way I'm bundling it with `esbuild` might be the problem.

**Comment by rpedrosanto** [comment]
It turns out that the problem is I have a web worker and I was losing the types during serialization when communicating with it.
Thanks for you attention @msujew !

**Reply by msujew** [reply, marked as answer]
I see, we used to have a similar issue. We addressed this by adding some more functionality to the `JsonSerializer` service. Take a look at [the serializer options interface](https://github.com/langium/langium/blob/0b603a61b4725bc74adf6c21adcff586e129e1e1/packages/langium/src/serializer/json-serializer.ts#L15-L21).

**Reply by luiz290788** [reply]
@msujew Do you have an example on how to do that? I was able to access `node.$cstNode?._text` but then I get a typescript error.

**Reply by msujew** [reply]
@luiz290788 What exactly are you trying to do? Can't you just use `node.$cstNode?.text`? `_text` is specifically not accessible.

Note that this question in particular was more related to serialization of AST nodes.

**Reply by luiz290788** [reply]
I have a command that returns the AST. I'm calling that command from the main thread of the browser. The command handler is getting the document and returning as the result of the command. When I get the returned value from the command `node.$cstNode.text` is not defined because it was lost in the serialization.

I was able to avoid the issue in my solution. I won't need to use the serializer, but it would be nice to learn about that. It also looks like that the serializer is not accessible from the command handler because it is not part of the shared services.

**Reply by msujew** [reply]
> It also looks like that the serializer is not accessible from the command handler because it is not part of the shared services.

Right, but you can put the `document.uri` into the `ServiceRegistry` to find the serializer that belongs to the document that you're trying to serialize.
